src/boss/worker.go

```diff
@@ -2,214 +2,479 @@ package boss
 
 import (
 	"fmt"
+	"io"
 	"log"
-	"os"
-	"path/filepath"
-	"strings"
 	"net/http"
+	"os"
+	"os/exec"
+	"os/user"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type WorkerState int
+
+const (
+	STARTING WorkerState = iota
+	RUNNING
+	CLEANING // waiting for already-started requests to complete (so can kill cleanly)
+	DESTROYING
 )
 
-type Invocation struct {
-	w    http.ResponseWriter
-	r    *http.Request
-	Done chan bool
+var (
+	clusterLogFile *os.File
+	taskLogFile    *os.File
+	clusterLog     *log.Logger
+	taskLog        *log.Logger
+	totalTask      int32
+	sumLatency     int64
+	nLatency       int64
+)
+
+type WorkerPool struct {
+	nextId  int
+	target  int
+	workers []map[string]*Worker
+	queue   chan *Worker
+	WorkerPoolPlatform
+	Scaling
+	sync.Mutex
 }
 
-func NewInvocation(w http.ResponseWriter, r *http.Request) *Invocation {
-	return &Invocation{w: w, r: r, Done: make(chan bool)}
+//platform specific attributes and functions
+type WorkerPoolPlatform interface {
+	NewWorker(workerId string) *Worker  //return new worker struct
+	CreateInstance(worker *Worker) //create new instance in the cloud platform
```

Comment:
This interface is used for platform-specific WorkerPool. For example, AzureWorkerPool and GcpWorkerPool inherits this interface. In this way, we can have a non platform-specific WorkerPool struct in worker.go file. But this design isn't good because we need to call AzureWorkerPool.WorkerPool.method() to call methods like "Close()". In other words, we are calling "up" in a class that has a lower inheritance level. We will refactor this part to make our code only call "down". That is, only classes that has a higher inheritance level can call methods/fields in its children classes.